delphi/GrFN/networks.py

```diff
@@ -0,0 +1,383 @@
+import xml.etree.ElementTree as ET
+from copy import deepcopy
+from pathlib import Path
+from typing import Dict
+from enum import Enum
+import importlib
+import inspect
+import time
+import json
+import os
+import ast
+
+import networkx as nx
+
+from delphi.translators.for2py import (
+    preprocessor,
+    translate,
+    get_comments,
+    pyTranslate,
+    genPGM,
+)
+
+import subprocess as sp
+
+
+def timeit(method):
+    """Timing wrapper for exectuion comparison."""
+    def timed(*args, **kw):
+        ts = time.time()
+        result = method(*args)
+        te = time.time()
+        print(f"{method.__name__}:\t{((te - ts) * 1000):2.4f}ms")
+        return result
+
+    return timed
+
+
+class GroundedFunctionNetwork(nx.DiGraph):
+    """
+    Stub for the Grounded Function Network class
+    """
+
+    def __init__(self, nodes, edges, subgraphs):
+        super(GroundedFunctionNetwork, self).__init__()
+        self.add_nodes_from(nodes)
+        self.add_edges_from(edges)
+        self.scopes = subgraphs
+
+        self.inputs = [n for n, d in self.in_degree() if d == 0]
+        self.outputs = [n for n, d in self.out_degree() if d == 0]
+        self.output_node = self.outputs[-1]
+
+    def __repr__(self):
+        return self.__str__()
+
+    def __str__(self):
+        return self.traverse_nodes(self.inputs)
+
+    def traverse_nodes(self, node_set, depth=0):
+        """BFS traversal of nodes that returns name traversal as large string."""
+        tab = "  "
+        result = ""
```

Comment:
Construct a list of strings and then concatenate the list, instead of doing `+=` (see https://docs.python-guide.org/writing/structure/#mutable-and-immutable-types).